llvm_mode LTO instrumentlist and persistent tests fails on Alpine Linux riscv64

**IMPORTANT**

AFLplusplus-4.20c


**Describe the bug**
On alpine linux, riscv64 the LTO llvm_mode test fails.

```
[*] Testing: LTO llvm_mode
[!] LTO llvm_mode failed   
[!] llvm_mode LTO instrumentlist feature compilation failed
[!] llvm_mode LTO persistent mode feature compilation failed
```

**To Reproduce**

On an Alpin eLinux riscv64 machine run:
```
        export PATH="$PATH:/usr/lib/llvm17/bin"
        export NO_NYX=1
        export AFL_NO_X86=1
        export LLVM_LTO=1
        make PREFIX=/usr all

        CFLAGS= CXXFLAGS= make AFL_NO_X86=1 test
```

**Expected behavior**
Test passes

**Screen output/Screenshots**

```
[*] Testing: LTO llvm_mode
[!] LTO llvm_mode failed
[!] llvm_mode LTO instrumentlist feature compilation failed
[!] llvm_mode LTO persistent mode feature compilation failed
```

**Additional context**

This was discovered on riscv64.

> does this work? `clang -flto -o test test-instr.c`

No because there exist a `test` directory. But I if output it to `test1` it does work.
```
ncopa-edge-riscv64:~/aports/community/afl++/src/AFLplusplus-4.20c$ clang -flto -o test test-instr.c 
/usr/bin/ld: cannot open output file test: Is a directory
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ncopa-edge-riscv64:~/aports/community/afl++/src/AFLplusplus-4.20c$ clang -flto -o test1 test-instr.c
ncopa-edge-riscv64:~/aports/community/afl++/src/AFLplusplus-4.20c$
```

There is also a `utils/persistent_mode/test-instr.c`:
```
ncopa-edge-riscv64:~/aports/community/afl++/src/AFLplusplus-4.20c$ cd utils/persistent_mode/
ncopa-edge-riscv64:~/aports/community/afl++/src/AFLplusplus-4.20c/utils/persistent_mode$ clang -flto -o test test-instr.c 
test-instr.c:22:1: error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Wimplicit-int]
   22 | __AFL_FUZZ_INIT();
      | ^
      | int
test-instr.c:31:3: error: call to undeclared function '__AFL_INIT'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   31 |   __AFL_INIT();
      |   ^
test-instr.c:32:24: error: use of undeclared identifier '__AFL_FUZZ_TESTCASE_BUF'
   32 |   unsigned char *buf = __AFL_FUZZ_TESTCASE_BUF;
      |                        ^
test-instr.c:34:10: error: call to undeclared function '__AFL_LOOP'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   34 |   while (__AFL_LOOP(UINT_MAX)) {  // if you have 100% stability
      |          ^
test-instr.c:36:24: error: use of undeclared identifier '__AFL_FUZZ_TESTCASE_LEN'
   36 |     unsigned int len = __AFL_FUZZ_TESTCASE_LEN;
      |                        ^
```

But I assume you meant the toplevel file.

--- a/GNUmakefile
+++ b/GNUmakefile
@@ -369,9 +369,18 @@ man:    $(MANPAGES)
 test:	tests
 
 .PHONY: tests
+ifdef ARCH
+tests:
+	@command -v docker >/dev/null 2>&1 || { echo "[-] Error: docker is required for ARCH testing"; exit 1; }
+	@echo "[*] Running tests in $(ARCH) Docker container via QEMU..."
+	@echo "[*] Output logged to test/test-$(ARCH).log"
+	@docker run --rm --privileged multiarch/qemu-user-static --reset -p yes > /dev/null 2>&1 || true
+	docker build --progress=plain --platform linux/$(ARCH) -f test/Dockerfile.qemu -t aflpp-test-$(ARCH) . 2>&1 | tee test/test-$(ARCH).log
+else
 tests:	source-only binary-only
 	@cd test ; ./test-all.sh
 	@rm -f test/errors
+endif
 
 .PHONY: performance-tests
 performance-tests:	performance-test

--- a/test/Dockerfile.qemu
+++ b/test/Dockerfile.qemu
@@ -1,0 +1,16 @@
+FROM alpine:latest AS base
+
+RUN apk add --no-cache \
+    build-base clang clang-dev lld llvm-dev llvm-static \
+    python3 bash coreutils linux-headers
+
+ENV NO_NYX=1 AFL_NO_X86=1 LLVM_LTO=1 \
+    AFL_SKIP_CPUFREQ=1 AFL_TRY_AFFINITY=1 \
+    AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES=1
+
+FROM base
+WORKDIR /AFLplusplus
+COPY . .
+
+RUN make -j$(nproc) source-only
+RUN cd test && ./test-all.sh

--- a/test/test-llvm-lto.sh
+++ b/test/test-llvm-lto.sh
@@ -14,7 +14,7 @@ test -e ../afl-clang-lto -a -e ../SanitizerCoverageLTO.so && {
   }
 
   rm -f test-instr.plain
-  ../afl-clang-lto -o test-instr.plain ../test-instr.c > /dev/null 2>&1
+  AFL_DEBUG=1 ../afl-clang-lto -o test-instr.plain ../test-instr.c > test-lto.out 2>&1
   test -e test-instr.plain && {
     chmod +x test-instr.plain
     ls -l test-instr.plain
@@ -41,10 +41,13 @@ test -e ../afl-clang-lto -a -e ../SanitizerCoverageLTO.so && {
     }
     rm -f test-instr.plain.0 test-instr.plain.1
   } || {
+    echo CUT------------------------------------------------------------------CUT
+    cat test-lto.out
+    echo CUT------------------------------------------------------------------CUT
     $ECHO "$RED[!] LTO llvm_mode failed"
     CODE=1
   }
-  rm -f test-instr.plain
+  rm -f test-instr.plain test-lto.out
 
   echo foobar.c > instrumentlist.txt
   AFL_DEBUG=1 AFL_LLVM_INSTRUMENT_FILE=instrumentlist.txt ../afl-clang-lto -o test-compcov test-compcov.c > test.out 2>&1
@@ -59,11 +62,14 @@ test -e ../afl-clang-lto -a -e ../SanitizerCoverageLTO.so && {
       CODE=1
     }
   } || {
+    echo CUT------------------------------------------------------------------CUT
+    cat test.out
+    echo CUT------------------------------------------------------------------CUT
     $ECHO "$RED[!] llvm_mode LTO instrumentlist feature compilation failed"
     CODE=1
   }
   rm -f test-compcov test.out instrumentlist.txt
-  ../afl-clang-lto -o test-persistent ../utils/persistent_mode/persistent_demo.c > /dev/null 2>&1
+  AFL_DEBUG=1 ../afl-clang-lto -o test-persistent ../utils/persistent_mode/persistent_demo.c > test-lto.out 2>&1
   test -e test-persistent && {
     echo foo | AFL_QUIET=1 ../afl-showmap -m none -o /dev/null -q -r ./test-persistent && {
       $ECHO "$GREEN[+] llvm_mode LTO persistent mode feature works correctly"
@@ -72,10 +78,13 @@ test -e ../afl-clang-lto -a -e ../SanitizerCoverageLTO.so && {
       CODE=1
     }
   } || {
+    echo CUT------------------------------------------------------------------CUT
+    cat test-lto.out
+    echo CUT------------------------------------------------------------------CUT
     $ECHO "$RED[!] llvm_mode LTO persistent mode feature compilation failed"
     CODE=1
   }
-  rm -f test-persistent
+  rm -f test-persistent test-lto.out
 } || {
   $ECHO "$YELLOW[-] LTO llvm_mode not compiled, cannot test"
   INCOMPLETE=1

--- a/test/test-llvm.sh
+++ b/test/test-llvm.sh
@@ -165,7 +165,7 @@ test -e ../afl-clang-fast -a -e ../split-switches-pass.so && {
               CODE=1
               ;;
         esac
-        rm -f in2/in*
+        rm -rf in2
       }
       export AFL_QUIET=1
       if type bash >/dev/null ; then {
@@ -221,7 +221,7 @@ test -e ../afl-clang-fast -a -e ../split-switches-pass.so && {
 
   AFL_LLVM_INSTRUMENT=AFL AFL_DEBUG=1 AFL_LLVM_LAF_SPLIT_SWITCHES=1 AFL_LLVM_LAF_TRANSFORM_COMPARES=1 AFL_LLVM_LAF_SPLIT_COMPARES=1 ../afl-clang-fast -o test-compcov.compcov test-compcov.c > test.out 2>&1
   test -e test-compcov.compcov && test_compcov_binary_functionality ./test-compcov.compcov && {
-    grep --binary-files=text -Eq " [ 123][0-9][0-9] location| [3-9][0-9] location" test.out && {
+    grep $GREPAOPTION -Eq " [ 123][0-9][0-9] location| [3-9][0-9] location" test.out && {
       $ECHO "$GREEN[+] llvm_mode laf-intel/compcov feature works correctly"
     } || {
       $ECHO "$RED[!] llvm_mode laf-intel/compcov feature failed"

